### ISSUE TITLE
layer_norm_backward assumes 2-D input: silently wrong gradients + out-of-bounds writes for rank ≥ 3 inputs

### ISSUE THREAD

**Issue by ethanZZZZZZZZZZZZ** [FlagOpen/FlagGems#5392]
### Describe the bug

# `layer_norm_backward` assumes 2-D input: silently wrong gradients + out-of-bounds writes for rank ≥ 3 inputs

## Summary

`layer_norm_forward` supports inputs of arbitrary rank (it derives `N` from `normalized_shape`), but `layer_norm_backward` hard-codes a 2-D view of the input. For the dominant real-world case — `[batch, seq, hidden]` LayerNorm in transformers — the backward pass (a) computes **numerically wrong gradients for dx, dW and dB**, and (b) makes `weight_bias_backward_kernel` **write out of bounds** past the `weight_grad`/`bias_grad` buffers, corrupting the CUDA caching-allocator heap. Depending on the heap layout this manifests as illegal-memory-access crashes in *unrelated* kernels later in the process, or — worse — no crash at all, just silently wrong gradients.

The bug is pure shape arithmetic, so it is platform-independent (reproducible on CUDA and other backends). It is present in the 5.0.2 release and still present on `master`.

## Root cause

`src/flag_gems/ops/layernorm.py`:

```python
# forward (lines 350-351 on master) — correct: N comes from normalized_shape
N = math.prod(normalized_shape)
M = input.numel() // N

# backward (lines 431-432 on master) — wrong: assumes input is 2-D
M = input.shape[0]
N = input.numel() // M
```

For `input.shape = [52, 7, 64]` with `normalized_shape = (64,)`, the correct split is `M = 364, N = 64`, but backward computes `M = 52, N = 448`. Consequences:

1. **Wrong values.** `layer_norm_backward_kernel` reduces over 448-wide "rows", indexes `mean`/`rstd` with the wrong row count, and loads `W` at `cols < N` offsets — `mean`, `rstd`, and `W` only have 364 / 364 / 64 elements respectively.
2. **Out-of-bounds writes.** `weight_grad`/`bias_grad` are allocated as `torch.empty_like(weight)` (64 floats), but `weight_bias_backward_kernel` stores at `pid < N` (up to 447):

   ```python
   col_mask = pid < N                     # pid up to 447
   tl.store(dW + pid, dw, mask=col_mask)  # dW has only 64 elements -> OOB write
   tl.store(dB + pid, db, mask=col_mask)
   ```

   With realistic shapes this clobbers neighboring caching-allocator blocks; with larger shapes it faults deterministically (observed on MetaX MACA as an ATU translation fault naming `layer_norm_backward_kernel`; on CUDA this should surface as an illegal memory access).

Note that `M * N == numel` under both splits, so most buffer accesses stay in-bounds — which is exactly why the bug usually **does not crash** and instead returns wrong gradients silently.

## Minimal reproduction (pure FlagGems, no external plugins)

```python
import torch
import flag_gems

torch.manual_seed(0)
M3, L, H = 52, 7, 64

x_cpu = torch.randn(M3, L, H)
w_cpu = torch.randn(H)
b_cpu = torch.randn(H)
g_cpu = torch.randn(M3, L, H)

# aten CPU reference
xr = x_cpu.clone().requires_grad_(True)
wr = w_cpu.clone().requires_grad_(True)
br = b_cpu.clone().requires_grad_(True)
torch.nn.functional.layer_norm(xr, (H,), wr, br).backward(g_cpu)

# FlagGems on GPU
x = x_cpu.cuda().requires_grad_(True)
w = w_cpu.cuda().requires_grad_(True)
b = b_cpu.cuda().requires_grad_(True)
with flag_gems.use_gems():
    torch.nn.functional.layer_norm(x, (H,), w, b).backward(g_cpu.cuda())

print("gx maxdiff:", float((x.grad.cpu() - xr.grad).abs().max()))
print("gw maxdiff:", float((w.grad.cpu() - wr.grad).abs().max()))
print("gb maxdiff:", float((b.grad.cpu() - br.grad).abs().max()))
```

Observed (flag_gems 5.0.2, triton 3.6.0, torch 2.10):

```
gx maxdiff: 12.872157096862793      # expect ~1e-6
gw maxdiff: 48.97093200683594       # expect ~1e-5
gb maxdiff: 47.84791183471680       # expect ~1e-5
```

Deterministic OOB variant — change the shape to `M3, L, H = 4, 65536, 64`: the mis-computed `N` becomes 4.2M, and `layer_norm_backward_kernel` / `weight_bias_backward_kernel` read/write megabytes past the 64-element `W`/`dW`/`dB` buffers, producing a device-side illegal-address fault.

## Suggested fix

Two lines, aligning backward with forward:

```diff
--- a/src/flag_gems/ops/layernorm.py
+++ b/src/flag_gems/ops/layernorm.py
@@ def layer_norm_backward(grad_out, input, normalized_shape, mean, rstd, ...):
-    M = input.shape[0]
-    N = input.numel() // M
+    N = math.prod(normalized_shape)
+    M = input.numel() // N
```

No kernel or launch-site changes are needed: `M * N == numel` still holds, and `mean`/`rstd` have exactly `M` elements under the corrected split. For genuine 2-D inputs, `math.prod(normalized_shape) == input.shape[1]`, so behavior is unchanged.

With this patch applied locally, the reproduction above reports:

```
gx maxdiff: 9.5367431640625e-07
gw maxdiff: 7.62939453125e-06
gb maxdiff: 1.52587890625e-05
```

and the amplified-shape variant completes without any device fault. We also re-ran the Hugging Face `all-mpnet-base-v2` test suite (`test_modeling_mpnet.py`, 196 tests) through a FlagGems-routed backend where this bug previously caused a heap-corruption cascade; the previously failing training test now passes and no device faults remain.

## Environment

- flag_gems 5.0.2 (PyPI wheel); bug also confirmed by code inspection on `master` (lines 431-432)
- triton 3.6.0, torch 2.10
- Reproduced on MetaX MACA; the defect is shape arithmetic and therefore expected to reproduce identically on CUDA
